File: Frontend/src/domain/courses/types/ManagedCoursePageDto.ts
import type { CategoryDto } from "./CourseSummaryDto";
import type { CourseStructureDtoApi } from "./CoursePageDto";
import type { Money } from "./money";
import type {
  ManagedCourseLinks,
  ManagedModuleLinks,
  ManagedLessonLinks,
} from "./links";

/** Managed course page: course.data (CourseCoreDto) */
export interface CourseCoreDto {
  id: string;
  title: string | null;
  description: string | null;
  status: "Draft" | "Published" | "Deleted";
  price: Money;
  updatedAtUtc: string;
  imageUrls: string[] | null;
  tags: string[] | null;
  instructorId: string;
  categoryId: string;
}

export interface ManagedCoursePageCourseDto {
  data: CourseCoreDto;
  links: ManagedCourseLinks;
}

/** Managed module: data */
export interface ModuleCoreDto {
  id: string;
  title: string | null;
  lessonCount: number;
  duration: string;
}

export interface ManagedCoursePageModuleDto {
  data: ModuleCoreDto;
  links: ManagedModuleLinks;
}

/** Managed lesson: data (LessonCoreDto) */
export interface LessonCoreDto {
  id: string;
  title: string | null;
  index: number;
  duration: string;
  thumbnailUrl: string | null;
  access: "Private" | "Public";
  moduleId: string;
  courseId: string;
  description: string | null;
  videoUrl: string | null;
  transcriptUrl: string | null;
}

export interface ManagedCoursePageLessonDto {
  data: LessonCoreDto;
  links: ManagedLessonLinks;
}

/**
 * Backend DTO: GET /manage/courses/{id} (strongly-typed links).
 */
export interface ManagedCoursePageDto {
  course: ManagedCoursePageCourseDto;
  structure: CourseStructureDtoApi;
  modules: Record<string, ManagedCoursePageModuleDto> | null;
  lessons: Record<string, ManagedCoursePageLessonDto> | null;
  categories: Record<string, CategoryDto> | null;
}

File: src/CourseService/Application/Pages/ManagedCoursePage/ManagedCoursePageDto.cs
using Courses.Application.Categories.Dtos;
using Courses.Application.Courses.Dtos;
using Courses.Application.Lessons.Dtos;
using Courses.Application.Modules.Dtos;
using Courses.Application.Pages.Dtos;

namespace Courses.Application.Pages.ManagedCoursePage;

public record ManagedCoursePageDto
{
    public required CourseDto Course { get; init; }
    public required CourseStructureDto Structure { get; init; }

    public required IReadOnlyDictionary<Guid, ManagedModuleDto> Modules { get; init; }
    public required IReadOnlyDictionary<Guid, LessonDto> Lessons { get; init; }
    public required IReadOnlyDictionary<Guid, CategoryDto> Categories { get; init; }
}
